packages/flutter/lib/src/widgets/form.dart

```diff
@@ -56,8 +65,23 @@ class FormState extends State<Form> {
   int _generation = 0;
   Set<FormFieldState<dynamic>> _fields = new Set<FormFieldState<dynamic>>();
 
-  /// Called when a form field has changed. This will cause all form fields
-  /// to rebuild, useful if form fields have interdependencies.
+  @override
+  void dependenciesChanged() {
+    super.dependenciesChanged();
+    final ModalRoute<Null> route = ModalRoute.of(context);
+    if (route != null && route.isCurrent)
+      route.scopedWillPopCallback = config.onWillPop;
```

Comment:
You're right (Ian had brought this up as well, but I thought I could avoid it). A Route can have an arbitrary number of Form descendants.  So I'll need to manage a list of callbacks, rather than just one.